github_server/action.yml

```diff
@@ -0,0 +1,53 @@
+name: "ALKiln/github_server: Isolated docassemble tests"
+description: "Automatically test docassemble interviews on GitHub's servers, in an isolated environment. Avoids docassemble reload errors and other external factors"
+
+
+inputs:
+  CONFIG_CONTENTS:
+    description: "The contents of the config file of the docassemble server"
+    required: false
+  # This is necessary to test pre-releases at the very least. It does seem
+  # strange considering we're controlling everything through the action, but
+  # that actually doesn't touch the npm version that's pulled in.
+  ALKILN_VERSION:
+    description: 'Version of ALKiln to get from npm.'
+    required: false
+    default: '^5.0.0'
+  MAX_SECONDS_FOR_SETUP:
+    description: "Amount of time to give the Project to upload your package's GitHub code."
+    required: false
+    default: '120'
+  SERVER_RELOAD_TIMEOUT_SECONDS:
+    description: "Max amount of time to give the server to reload if it needs to"
+    required: false
+    # no default to let the custom Scenario timeouts create the reload timeout
+
+
+runs:
+  using: "composite"
+  steps:
+      - name: "github_server: Set environment variables"
+        run: |
+          echo "DA_ADMIN_API_KEY=abcd1234abcd1234abcd5678abdc5678" >> $GITHUB_ENV
+        shell: bash
+      - name: Download and start docker
+        run: |
+          mkdir /tmp/config
+          echo "$CONFIG_CONTENTS" > /tmp/config/myconfig.yml 
+          docker pull brycewilley/smoldocassemble
+          docker run --env DA_ADMIN_EMAIL="admin@example.com" --env DA_ADMIN_PASSWORD="password" --env DA_ADMIN_API_KEY="${{ env.DA_ADMIN_API_KEY }}" --env DA_CONFIG=/tmp/config/myconfig.yml --volume /tmp/config:/tmp/config --cap-add SYS_PTRACE --memory="4gb" -d -p 80:80 jhpyle/docassemble
+          # TODO before merge: do a while loop here until `localhost` shows "It's ready"
+          sleep 480
+          curl localhost
+        shell: bash
+      - name: Run Kiln
+        # TODO: Research how to get this to always be the same branch we're on
+        uses: SuffolkLITLab/ALKiln@v5
+        with:
+          SERVER_URL: "http://localhost"
+          DOCASSEMBLE_DEVELOPER_API_KEY: "${{ env.DA_ADMIN_API_KEY }}"
+      - name: Docker logs
+        # TODO: create an arifact for this instead
+        run: |
+          docker logs $(docker ps --filter "ancestor=jhpyle/docassemble" --format {{.ID}})
+        shell: bash
```

Comment:
Another refactor: the logs could have a few issues:

* this won't print if the ALKiln tests or the previous step fails for any reason. it should print in those cases too
* there is the chance that it prints secrets from the config file. The way that we set the config file is through the entire thing, which means I think that GitHub will only censor things if it prints the config in it's entirety, and won't censor specific lines. Some ways around that include:
  * making the config file a mako template or something similar, and replacing the actual secret values with specific github secrets. That would mean that github would still censor the actual secret parts in case they're printed in the log.
  * seeing if GitHub lets us restrict who can download artifact files, and making this a downloadable file instead of printing straight to the GH log?

Reply:
I've at least confirmed that a user who isn't logged in can't download the artifacts or see the details of the console log. I assume you mean restricting on a more fine-grained basis. E.g. only allow admins to see those.

Reply:
This is what I'm told about config logs:

> The configuration isn't echoed to output, but error messages might contain references to configuration values.

Reply:
Another alternative (not that I'm a huge fan) is to encrypt the output file. It's a lot more work for the user, but so far I haven't found a way to restrict permissions.

As for Mako + secrets, just to state it explicitly, that's up to the user to do in their own repository. I'm not sure what those files would look like and how you'd run those scripts. That, too, would be a lot of work for the user. Maybe more work than an encrypted file.